File: racket/oo/puzzle/self_dispatch.py
class c1:
    def m1(self):
        print "c1:m1"

    def m2(self):
        self.m1() # python use dynamic dispatch for self!


class c2(c1):
    def m1(self):
        print "c2:m1"


v = c2()
v.m2()

File: racket/oo/puzzle/self_dispatch.cc
#include <iostream>

using namespace std;

class c1 {
 public:
  void m1() {
    cout << "c1:m1" << endl;
  }
  void m2() {
    this->m1(); // C++ use static dispatch for self!
  }
};

class c2:public c1 {
 public:
  void m1() {
    cout << "c2:m1" << endl;
  }
};

int main(int argc, const char *argv[])
{
  c2* p = new c2();
  p->m2();
  return 0;
}
